### PR TITLE
chore: remove rbac psuedo resources, add custom verbs

### PR DESCRIPTION
**Pull request by Emyrk**
Supports https://github.com/coder/coder/issues/13226

# What this does

Removes our pseudo rbac resources like `WorkspaceApplicationConnect` in favor of additional verbs like `ssh`. This is to make more intuitive permissions for building custom roles.

Prior to this change, trying to create a custom role creator would not be intuitive or dynamic.

Screenshot from 2024-05-14 19-07-03

# Implementation

The source of truth is now [`policy.go` ](https://github.com/coder/coder/blob/2255162cb546992f6728f17b19e0cfa1f9ba234e/coderd/rbac/policy/policy.go#L63-L63).

To prevent even more line changes by drastically changing the golang code, the previous resource list is autogenerated from this source of truth: https://github.com/coder/coder/blob/2255162cb546992f6728f17b19e0cfa1f9ba234e/coderd/rbac/object_gen.go#L218-L220

The autogen is the easiest way to ensure consistency while still using the `rbac.Object`.

The codersdk authz constants are now also in sync via auto gen: https://github.com/coder/coder/blob/eeb6ac20988b77ae7bf577c79b304fa34a7848f5/codersdk/rbacresources_gen.go

# Additional checks

The prior and current implementation do not force the caller to populate the object fields (owner, org, id, acls, etc) or use a correct action when calling the rego.

Now that we have a definitive policy location, we can add at least runtime checks for correctness. I have only added correctness checks for actions on unit tests.

https://github.com/coder/coder/blob/d60f65ac83141bdd34ac7f0c1d01041ebb4de76c/coderd/rbac/authz.go#L243-L247

This will prevent asserting actions that are not available for a given object.

### CODE REVIEW

**Comment by Emyrk** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/13276?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#13276** <a href="https://app.graphite.dev/github/pr/coder/coder/13276?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#13274** <a href="https://app.graphite.dev/github/pr/coder/coder/13274?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Emyrk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/rbac/policy/policy.go [hunk 1]
```diff
@@ -8,4 +14,264 @@ const (
...
+	"user": {
+		Actions: map[Action]ActionDefinition{
+			// Actions deal with site wide user objects.
+			ActionRead:   actDef(0, "read user data"),
+			ActionCreate: actDef(0, "create a new user"),
+			ActionUpdate: actDef(0, "update an existing user"),
+			ActionDelete: actDef(0, "delete an existing user"),
+
+			ActionReadPersonal:   actDef(fieldOwner, "read personal user data like password"),
+			ActionUpdatePersonal: actDef(fieldOwner, "update personal data"),
```
Comment (Emyrk) [inline coderd/rbac/policy/policy.go:103]: This used to be `ResourceUserData`, but is now just additional verbs

File: Makefile [hunk 2]
```diff
@@ -611,7 +612,10 @@ examples/examples.gen.json: scripts/examplegen/main.go examples/examples.go $(sh
 	go run ./scripts/examplegen/main.go > examples/examples.gen.json
 
 coderd/rbac/object_gen.go: scripts/rbacgen/main.go coderd/rbac/object.go
-	go run scripts/rbacgen/main.go ./coderd/rbac > coderd/rbac/object_gen.go
+	go run scripts/rbacgen/main.go rbac > coderd/rbac/object_gen.go
+
+codersdk/rbacresources_gen.go: scripts/rbacgen/main.go coderd/rbac/object.go
+	go run scripts/rbacgen/main.go codersdk > codersdk/rbacresources_gen.go
```
Comment (Emyrk) [inline Makefile:618]: Before this was manually _kept in sync_. Or so we told ourselves

File: coderd/database/dbauthz/dbauthz.go [hunk 3]
```diff
@@ -164,14 +164,13 @@ var (
...
-					rbac.ResourceWorkspaceBuild.Type: {policy.ActionRead, policy.ActionUpdate, policy.ActionDelete},
-					rbac.ResourceUserData.Type:       {policy.ActionRead, policy.ActionUpdate},
-					rbac.ResourceAPIKey.Type:         {rbac.WildcardSymbol},
+					rbac.ResourceFile.Type:     {policy.ActionRead},
+					rbac.ResourceSystem.Type:   {rbac.WildcardSymbol},
+					rbac.ResourceTemplate.Type: {policy.ActionRead, policy.ActionUpdate},
+					// Unsure why provisionerd needs update and read personal
+					rbac.ResourceUser.Type:      {policy.ActionRead, policy.ActionReadPersonal, policy.ActionUpdatePersonal},
+					rbac.ResourceWorkspace.Type: {policy.ActionRead, policy.ActionUpdate, policy.ActionDelete, policy.ActionWorkspaceBuild},
+					rbac.ResourceApiKey.Type:    {rbac.WildcardSymbol},
```
Comment (Emyrk) [inline coderd/database/dbauthz/dbauthz.go:173]: `ResourceWorkspaceBuild` -> `ResourceWorkspace: ActionWorkspaceBuild`

(unsure why it needs this, keeping status quo)
`ResourceUserData` -> `ResourceUser: ActionReadPersonal, ActionUpdatePersonal`

File: coderd/database/dbauthz/dbauthz.go [hunk 4]
```diff
@@ -192,11 +191,10 @@ var (
...
 				Site: rbac.Permissions(map[string][]policy.Action{
-					rbac.ResourceSystem.Type:         {rbac.WildcardSymbol},
-					rbac.ResourceTemplate.Type:       {policy.ActionRead, policy.ActionUpdate},
-					rbac.ResourceWorkspace.Type:      {policy.ActionRead, policy.ActionUpdate},
-					rbac.ResourceWorkspaceBuild.Type: {policy.ActionRead, policy.ActionUpdate, policy.ActionDelete},
-					rbac.ResourceUser.Type:           {policy.ActionRead},
+					rbac.ResourceSystem.Type:    {rbac.WildcardSymbol},
+					rbac.ResourceTemplate.Type:  {policy.ActionRead, policy.ActionUpdate},
+					rbac.ResourceWorkspace.Type: {policy.ActionRead, policy.ActionUpdate, policy.ActionWorkspaceBuild},
+					rbac.ResourceUser.Type:      {policy.ActionRead},
```
Comment (Emyrk) [inline coderd/database/dbauthz/dbauthz.go:197]: `ResourceWorkspaceBuild` -> `ResourceWorkspace:  ActionWorkspaceBuild`

File: coderd/database/dbauthz/dbauthz.go [hunk 5]
```diff
@@ -234,19 +232,15 @@ var (
...
 					rbac.ResourceOrganizationMember.Type: {policy.ActionCreate},
-					rbac.ResourceOrgRoleAssignment.Type:  {policy.ActionCreate},
 					rbac.ResourceProvisionerDaemon.Type:  {policy.ActionCreate, policy.ActionUpdate},
-					rbac.ResourceUser.Type:               {policy.ActionCreate, policy.ActionUpdate, policy.ActionDelete},
-					rbac.ResourceUserData.Type:           {policy.ActionCreate, policy.ActionUpdate},
-					rbac.ResourceWorkspace.Type:          {policy.ActionUpdate},
-					rbac.ResourceWorkspaceBuild.Type:     {policy.ActionUpdate},
-					rbac.ResourceWorkspaceExecution.Type: {policy.ActionCreate},
+					rbac.ResourceUser.Type:               rbac.ResourceUser.AvailableActions(),
+					rbac.ResourceWorkspace.Type:          {policy.ActionUpdate, policy.ActionDelete, policy.ActionWorkspaceBuild, policy.ActionSSH},
```
Comment (Emyrk) [inline coderd/database/dbauthz/dbauthz.go:243]: `ResourceWorkspaceExecution` -> `ResourceWorkspace: ActionSSH`

File: coderd/rbac/README.md [hunk 6]
```diff
@@ -106,7 +106,7 @@ You can test outside of golang by using the `opa` cli.
 
 **Evaluation**
 
-opa eval --format=pretty 'false' -d policy.rego -i input.json
+opa eval --format=pretty "data.authz.allow" -d policy.rego -i input.json
```
Comment (Emyrk) [inline coderd/rbac/README.md:109]: This was a typo I found in the readme.

File: coderd/rbac/policy/policy.go [hunk 7]
```diff
@@ -8,4 +10,231 @@ const (
...
+	// Workspace provisioning
+	ActionWorkspaceBuild: actDef("allows starting, stopping, and updating a workspace"),
+
+	// Running a workspace
+	ActionSSH:                actDef("ssh into a given workspace"),
+	ActionApplicationConnect: actDef("connect to workspace apps via browser"),
+}
+
+// RBACPermissions is indexed by the type
+var RBACPermissions = map[string]PermissionDefinition{
```
Comment (Emyrk) [inline coderd/rbac/policy/policy.go:63]: The new source of truth for our policies.

File: coderd/rbac/roles_test.go [hunk 8]
```diff
@@ -359,18 +550,32 @@ func TestRolePermissions(t *testing.T) {
...
+	if passed {
+		for rtype, v := range remainingPermissions {
+			// nolint:tparallel -- Making a subtest for easier diagnosing failures.
+			t.Run(fmt.Sprintf("%s-AllActions", rtype), func(t *testing.T) {
+				if len(v) > 0 {
+					assert.Equal(t, map[policy.Action]bool{}, v, "remaining permissions should be empty for type %q", rtype)
+				}
+			})
+		}
+	}
```
Comment (Emyrk) [inline coderd/rbac/roles_test.go:578]: I added this to force us to update this test if resources or actions are added. This is sort of our source of truth for our "policy".


Would be really awesome to come up with a custom syntax for making this more compressed...

**Review by johnstcn — APPROVED** [review]
👍

**Comment by Emyrk** [comment]
### Merge activity

* **May 15, 10:46 AM EDT**: @Emyrk started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/coder/coder/13276).

**Pull request merged**